src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.NativeAot.cs

```
@@ -0,0 +1,218 @@
+            {
+                if (!TryGetNativeReaderForBlob(module, ReflectionMapBlob.ExternalTypeMap, out NativeReader externalTypeMapReader))
+                {
+                    continue;
+                }
+                NativeParser externalTypeMapParser = new NativeParser(externalTypeMapReader, 0);
+                NativeHashtable externalTypeMapTable = new NativeHashtable(externalTypeMapParser);
+
+                ExternalReferencesTable externalReferences = default;
+                externalReferences.InitializeCommonFixupsTable(module);
+
+                var lookup = externalTypeMapTable.Lookup(typeMapGroupHandle.GetHashCode());
+                NativeParser entryParser;
+                while (!(entryParser = lookup.GetNext()).IsNull)
+                {
+                    RuntimeTypeHandle foundTypeMapGroup = externalReferences.GetRuntimeTypeHandleFromIndex(entryParser.GetUnsigned());
+                    if (!foundTypeMapGroup.Equals(typeMapGroupHandle))
+                    {
+                        continue;
+                    }
+                    bool isValid = entryParser.GetUnsigned() == 1;
+                    if (!isValid)
+                    {
+                        unsafe
+                        {
+                            delegate*<void> exceptionStub = (delegate*<void>)externalReferences.GetFunctionPointerFromIndex(entryParser.GetUnsigned());
+                            exceptionStub();
+                            Debug.Fail("Expected exception stub to throw an exception.");
+                            return null; // Should never reach here, as the exception stub should throw an exception.
+                        }
+                    }
+
+                    return new ExternalTypeMapDictionary(new NativeHashtable(entryParser), externalReferences);
+                }
+            }
+
+            return new Dictionary<string, Type>(); // Return an empty dictionary if no valid type map is found.
```

Random thought: if you only call these APIs with a valid type map group, does ILC trim this path out entirely? If not, would it make sense to just return another instance of `ExternalTypeMapDictionary` with no content, or even just some other dummy type implementing the interface, to avoid constructing that `Dictionary<string, Type>` generic type instantiation? In theory it might save a tiny bit of size in apps that don't use `Dictionary<__Canon, __Canon>` at all? (not sure if it's worth it, just figured I'd mention it).

---

Or, should the compiler produce empty maps as needed and should this throw missing metadata exception instead?